core/arch/interp.c

```diff
@@ -3377,13 +3381,15 @@ build_bb_ilist(dcontext_t *dcontext, build_bb_t *bb)
             bb->instr_start = bb->cur_pc;
             if (bb->full_decode) {
                 /* only going through this do loop once! */
-                bb->cur_pc = decode(dcontext, bb->cur_pc, bb->instr);
+                bb->cur_pc = IF_AARCH64_ELSE(decode_with_ldstex, decode)
+                    (dcontext, bb->cur_pc, bb->instr);
                 if (bb->record_translation)
                     instr_set_translation(bb->instr, bb->instr_start);
             } else {
                 /* must reset, may go through loop multiple times */
                 instr_reset(dcontext, bb->instr);
-                bb->cur_pc = decode_cti(dcontext, bb->cur_pc, bb->instr);
+                bb->cur_pc = IF_AARCH64_ELSE(decode_cti_with_ldstex, decode_cti)
```

Comment:
> > These hidden decode-ahead routines (hidden from bb building) are bypassing a lot of bb building logic and checks: code origin checks and state updates (whether should go onto a new page, whether we'll crash decoding onto the next page, tracking which pages a block covers for flushing purposes), max instr count checks (if near limit and then see ldxr this will blow past the limit)
> 
> How are they bypassing these checks? The block builder already handles variable-length X86 instructions, so it ought to be able to handle these variable-length macro-instructions. There's certainly a danger of bugs as we, in effect, move from constant-length to variable-length instructions on AArch64, but I don't see why there should be a problem in principle.

The difference is that ldxr..stxr sequences don't have to be straight-line code: they can contain an arbitrary number of jumps and theoretically could hop around and touch many different pages. All without calling any of the check_new_page_* routines that bb building assumes, skipping -max_elide_* checks (can live with, but should document), stopping point checks, etc.